chain/src/pipe.rs

```diff
@@ -167,15 +168,29 @@ fn validate_header(header: &BlockHeader, ctx: &mut BlockContext) -> Result<(), E
 }
 
 /// Fully validate the block content.
-fn validate_block(b: &Block, ctx: &mut BlockContext) -> Result<(), Error> {
-	if b.header.height > ctx.head.height + 1 {
+fn validate_block(block: &Block, ctx: &mut BlockContext) -> Result<(), Error> {
+	if block.header.height > ctx.head.height + 1 {
 		return Err(Error::Orphan);
 	}
 
 	let curve = secp::Secp256k1::with_caps(secp::ContextFlag::Commit);
-	try!(b.validate(&curve).map_err(&Error::InvalidBlockProof));
-
-	// TODO check every input exists as a UTXO using the UTXO index
+	try!(block.validate(&curve).map_err(&Error::InvalidBlockProof));
+
+	for input in &block.inputs {
+		// TODO check every input exists as a UTXO using the UTXO index
+
+		// now check that any coinbase inputs have matured sufficiently
+		if let Ok(output) = ctx.store.get_output_by_commit(&input.commitment()) {
```

Comment:
I'd reverse this condition and the next. Calling get_output_by_commit first incurs a db cost that's useless for non-coinbase.

Reply:
We need to call get_output_by_commit first, as we need the output to know if it is a coinbase output or not (unless I'm misunderstanding what you are suggesting).

Reply:
Yes, ignore me, mind fart :P